pyamg/strength.py

```diff
@@ -199,64 +202,41 @@ def classical_strength_of_connection(A, theta=0.1, block=None, norm='abs'):
         else:
             raise ValueError('Invalid choice of norm.')
 
+        # drop small numbers
         data[np.abs(data) < 1e-16] = 0.0
-        Sp = np.empty_like(A.indptr)
-        Sj = np.empty_like(A.indices)
-        Sx = np.empty_like(data)
-
-        if norm in ('abs', 'fro'):
-            amg_core.classical_strength_of_connection_abs(
-                N, theta, A.indptr, A.indices, data, Sp, Sj, Sx)
-        elif norm == 'min':
-            amg_core.classical_strength_of_connection_min(
-                N, theta, A.indptr, A.indices, data, Sp, Sj, Sx)
-        else:
-            raise ValueError('Unrecognized option for norm.')
-
-        # One pass through nnz to find largest entry, one to filter
-        S = sparse.csr_matrix((Sx, Sj, Sp), shape=[N, N])
-
-        # Take magnitude and scale by largest entry
-        S.data = np.abs(S.data)
-        S = scale_rows_by_largest_entry(S)
-        S.eliminate_zeros()
-
-    # SOC computed based on A as CSR
     else:
-        if not sparse.isspmatrix_csr(A):
-            warn('Implicit conversion of A to csr',
-                 sparse.SparseEfficiencyWarning)
-            A = sparse.csr_matrix(A)
-
-        Sp = np.empty_like(A.indptr)
-        Sj = np.empty_like(A.indices)
-        Sx = np.empty_like(A.data)
-
-        if norm == 'abs':
-            amg_core.classical_strength_of_connection_abs(
-                A.shape[0], theta, A.indptr, A.indices, A.data,
-                Sp, Sj, Sx)
-        elif norm == 'min':
-            amg_core.classical_strength_of_connection_min(
-                A.shape[0], theta, A.indptr, A.indices, A.data,
-                Sp, Sj, Sx)
-        else:
-            raise ValueError('Unknown norm')
+        data = A.data
+        N = A.shape[0]
+
+    Sp = np.empty_like(A.indptr)
+    Sj = np.empty_like(A.indices)
+    Sx = np.empty_like(data)
+
+    if norm in ('abs', 'fro'):
+        amg_core.classical_strength_of_connection_abs(
+            N, theta, A.indptr, A.indices, data, Sp, Sj, Sx)
+    elif norm == 'min':
+        amg_core.classical_strength_of_connection_min(
+            N, theta, A.indptr, A.indices, data, Sp, Sj, Sx)
+    else:
+        raise ValueError('Unrecognized option for norm for strength.')
 
-        S = sparse.csr_matrix((Sx, Sj, Sp), shape=A.shape)
+    S = sparse.csr_matrix((Sx, Sj, Sp), shape=[N, N])
 
-        if blocksize > 1:
-            S = amalgamate(S, blocksize)
+    # Take magnitude and scale by largest entry
+    S.data = np.abs(S.data)
+    S = scale_rows_by_largest_entry(S)
+    S.eliminate_zeros()
 
-            # Take magnitude and scale by largest entry
-            S.data = np.abs(S.data)
-            S = scale_rows_by_largest_entry(S)
-            S.eliminate_zeros()
+    if blocksize > 1 and not block:
+        S = amalgamate(S, blocksize)
+        warn('Implicit conversion of A to bsr', sparse.SparseEfficiencyWarning)
+        A = A.tobsr(blocksize=[blocksize,blocksize])
```

Comment:
@lukeolson @jbschroder check if this makes sense to you. The only way blocksize is > 1 and block=False is we pass in a BSR matrix and block=False. If we then amalgamate the SOC matrix, we are mapping it to the block-dimensions of the original matrix. For this to work with other routines, e.g., interp/restrict, we need to map A back to a BSR matrix.

Reply:
While `A` is converted within the routine, it's a copy (this is probably bad, but it's a different issue).  That last line `A = A.tobsr(blocksize=[blocksize,blocksize])` will create a new object that isn't returned.

So I think it's not needed? (I commented out that `tobsr()` line in the test below).  So I'd say go ahead.

```python
A = pyamg.gallery.poisson((4,)).tobsr(blocksize=(2,2))
print(f'{id(A)=}')

S = pyamg.strength.classical_strength_of_connection(A, block=True)
print(S.toarray())
print(A.toarray())
print(type(S))
print(type(A))

S = pyamg.strength.classical_strength_of_connection(A, block=False)
print(S.toarray())
print(A.toarray())
print(type(S))
print(type(A))
print(f'{id(A)=}')
```

output:
```
id(A)=4304610512
[[1.  0.5]
 [0.5 1. ]]
[[ 2. -1.  0.  0.]
 [-1.  2. -1.  0.]
 [ 0. -1.  2. -1.]
 [ 0.  0. -1.  2.]]
<class 'scipy.sparse._csr.csr_matrix'>
<class 'scipy.sparse._bsr.bsr_matrix'>

[[1. 1.]
 [1. 1.]]
[[ 2. -1.  0.  0.]
 [-1.  2. -1.  0.]
 [ 0. -1.  2. -1.]
 [ 0.  0. -1.  2.]]
<class 'scipy.sparse._csr.csr_matrix'>
<class 'scipy.sparse._bsr.bsr_matrix'>
id(A)=4304610512
```

Reply:
Hi All...just read through the conversation today.  The revised code here to just do 

    if blocksize > 1 and not block:
        S = amalgamate(S, blocksize)

    return S

to end the function makes sense.

@Ben @lukeolson   Is this PR ready to merge now?